missioncontrol_client/__init__.py

```diff
@@ -1,10 +1,83 @@
 import argparse
 import json
 import os
+
+# dl
+import io
+import zlib
+import socket
+import datetime
+from pyblake2 import blake2b
+from datalake import File
+
 from urllib.parse import urljoin
 
 import requests
 
+class UTC(object):
+    def __init__(self, d="now"):
+        if d == "now":
+            self._d = datetime.datetime.utcnow()
+        else:
+            raise NotImplementedError()
+
+    @property
+    def iso(self):
+        return self._d.isoformat() + "Z"
+
+
+class GzippedFileStream(File):
+
+    def __init__(self, fd, **metadata_fields):
+        self._left = ''
+        self._compressor = zlib.compressobj(2, zlib.DEFLATED, zlib.MAX_WBITS | 16)
+        super().__init__(fd, **metadata_fields)
+
+    def _initialize_methods_from_fd(self):
+        for m in ['seek', 'tell', 'close']:
+            setattr(self, m, getattr(self._fd, m))
+
+    def _read1(self, n=None):
+        while not self._left:
+            try:
+                self._left = next(self._fd)
+            except StopIteration:
+                break
+        ret = self._left[:n]
+        self._left = self._left[len(ret):]
+        return ret
+
+    def read(self, n=None):
+        l = bytearray()
+        if n is None or n < 0:
+            while True:
+                m = self._read1()
+                if not m:
+                    break
+                l.append(self._compressor.compress(m))
+        else:
+            while n > 0:
+                m = self._read1(n)
+                if not m:
+                    break
+                n -= len(m)
+                l.append(self._compressor.compress(m))
+        l.append(self._compressor.flush())
+        return b''.join(l)
+
+    def _calculate_hash(self):
+        '''ensure the hash is over the raw file, not the gzip steam'''
+        b2 = blake2b(digest_size=16)
```

Comment:
Is digest_size required here? Or just to ensure we're consistent?

Reply:
yeah, I we need both a hashing algorithm and a standard way to call it, otherwise we might have the same file twice with different hash lengths.
Actually, I'll probably encode the hash type and length with something like pymultihash.